src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs

```diff
@@ -160,9 +160,36 @@ private static ServiceContext CreateServiceContext(IOptions<KestrelServerOptions
 
                 async Task OnBind(ListenOptions options, CancellationToken onBindCancellationToken)
                 {
+                    var hasHttp1 = options.Protocols.HasFlag(HttpProtocols.Http1);
+                    var hasHttp2 = options.Protocols.HasFlag(HttpProtocols.Http2);
+                    var hasHttp3 = options.Protocols.HasFlag(HttpProtocols.Http3);
+                    var hasTls = options.IsTls;
+
+                    // Filter out invalid combinations.
+
+                    if (!hasTls)
+                    {
+                        // Http/1 without TLS, no-op HTTP/2 and 3.
+                        if (hasHttp1)
+                        {
+                            hasHttp2 = false;
+                            hasHttp3 = false;
+                        }
+                        // Http/3 requires TLS. Note we only let it fall back to HTTP/1, not HTTP/2
+                        else if (hasHttp3)
+                        {
+                            throw new InvalidOperationException("HTTP/3 requires UseHttps.");
```

Comment:
There are more ways to configure TLS than UseHttps? Make this message more general.

Reply:
Updated to "requires https".

I've added a test for ConfigureHttpsDefaults. I've also manually tested that quic works with the dev cert as well as being enabled from appsettings.json. I've added a config reader test, but it's not end-to-end, it's more about making sure the settings were correctly mapped to endpoitns.